### ISSUE TITLE
datadog agent 6.2.1 hangs on socket read with docker 1.12.1

### ISSUE THREAD

**Issue by jambii** [DataDog/datadog-agent#1856]
**Describe what happened:**
```
[root@redacted ~] sudo -u dd-agent -- datadog-agent check docker
=== Service Checks ===
[
  {
    "check": "docker.service_up",
    "host_name": "redacted",
    "timestamp": 1529361209,
    "status": 2,
    "message": "could not get docker containers: error listing containers: context deadline exceeded",
    "tags": null
  }
]
=========
Collector
=========

  Running Checks
  ==============
    docker
    ------
      Total Runs: 1
      Metrics: 0, Total Metrics: 0
      Events: 0, Total Events: 0
      Service Checks: 0, Total Service Checks: 0
      Average Execution Time : 5012ms
      Error: could not get docker containers: error listing containers: context deadline exceeded
      No traceback
      Warning: Error collecting containers: [could not get docker containers: error listing containers: context deadline exceeded]

```

strace of docker ps ... after read() of socket, no other calls succeed 
```
connect(3, {sa_family=AF_LOCAL, sun_path="/var/run/docker.sock"}, 23) = 0
epoll_create1(EPOLL_CLOEXEC)            = 4
epoll_ctl(4, EPOLL_CTL_ADD, 3, {EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, {u32=2144049400, u64=140250006132984}}) = 0
getsockname(3, {sa_family=AF_LOCAL, NULL}, [2]) = 0
getpeername(3, {sa_family=AF_LOCAL, sun_path="/var/run/docker.sock"}, [23]) = 0
futex(0xc82005c108, FUTEX_WAKE, 1)      = 1
read(3, 0xc82034a000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
futex(0x132cca8, FUTEX_WAIT, 0, NULL)   = 0
```

agent.log
```
Jun 18 15:30:24 redacted agent: 2018-06-18 15:30:24 PDT | ERROR | (docker.go:223 in Run) | Failed to inspect container 6a72d3cfdabe - context deadline exceeded
```

docker journal logs (repeats this pattern)
```
Jun 18 15:42:14 redacted docker[1100]: time="2018-06-18T15:42:14.138451715-07:00" level=error msg="Handler for GET /v1.24/containers/json returned error: write unix /var/run/docker.sock->@: write: broken pipe"
Jun 18 15:42:14 redacted docker[1100]: time="2018-06-18T15:42:14.138466374-07:00" level=error msg="Handler for GET /v1.24/containers/json returned error: write unix /var/run/docker.sock->@: write: broken pipe"
Jun 18 15:42:14 redacted docker[1100]: time="2018-06-18T15:42:14.138775435-07:00" level=error msg="Handler for GET /v1.24/containers/json returned error: write unix /var/run/docker.sock->@: write: broken pipe"
Jun 18 15:42:14 redacted dockerd[1100]: http: multiple response.WriteHeader calls
```

**Describe what you expected:**
I did not expect datadog to compromise docker. It's one thing if the check fails, but I was surprised to see that "external" docker commands started hanging too.
The original goal was to get container + host metrics... trying to move away from the deprecated docker plugin in 5.24.0  

specifically, had the following enabled: collect_container_size, collect_exit_codes, filtered_event_types (exec_start,exec_create_exec_die), collect_events

**Steps to reproduce the issue:**
Enable docker configuration as described above in datadog 6.2.1, using docker-engine-1.12.1 on cent7, 3.10 kernel. Restarted agent. After about 30 seconds, any attempt to run docker commands (e.g. info, ps) hangs

**Additional environment details (Operating System, Cloud provider, etc):**
run arguments (removed sensitive info): `root      1100  0.1  0.3 1682080 55676 ?       Ssl  Jun15   5:31 dockerd --graph=/data/docker -H tcp://x.x.x.x:2376 -H unix:///var/run/docker.sock --pidfile=/var/run/docker.pid --tlsverify=true`

package name: `docker-engine-1.12.1-1.el7.centos.x86_64`

build: `Linux redacted 3.10.0-862.3.3.el7.x86_64 #1 SMP Wed Jun 13 05:44:23 EDT 2018 x86_64 x86_64 x86_64 GNU/Linux`

tl;dr docker api weirdness after upgrade. is docker 1.12.1 supported?  I didn't have this issue with docker 17.0.4ce on a similarly configured system.

**Comment by xvello** [comment]
Hi @jambii 

Although docker 1.12.1 is EOL, we tested Agent6 with it to make sure it plays nice. My guess it that one of the options is putting a significant stress on the daemon, making it become unresponsive. Could you please try disabling options from the check to pinpoint if a particular one does? I'd start with `collect_container_size`, as this information is not kept by the docker daemon but recomputed (i.e `du -hs`) for every container, everytime.

Regards

**Comment by jambii** [comment]
Disabling collect_container_size fixed it. Looks like a bug with docker's devicemapper implementation. Thanks!
